torch/_dynamo/polyfills/pytree.py

```diff
@@ -95,3 +101,197 @@ def tree_leaves(
         return list(tree_iter(tree, is_leaf=is_leaf))
 
     __all__ += ["tree_leaves"]
+
+    class _Asterisk(str):
+        def __new__(cls) -> Self:
+            return super().__new__(cls, "*")
+
+        def __repr__(self) -> str:
+            return "*"  # no quotes
+
+    _asterisk = _Asterisk()
+    del _Asterisk
+
+    @dataclass(frozen=True)
+    class PyTreeSpec:
```

Comment:
@XuehaiPan, if I'm understanding this correctly, you're adding a polyfill for cxx_pytree.tree_flatten such that it will return an instance of this PyTreeSpec class. I'm not sure this works if we are trying to return a PyTreeSpec from a torch.compile'd function: does it create an instance of this "class PyTreeSpec" object, or does it create an instance of torch.utils._cxx_pytree.TreeSpec?

Reply:
In a compiled function, it returns `torch._dynamo.polyfills.pytree.PyTreeSpec`. This class provides the exactly same interfaces with `torch.utils._cxx_pytree.TreeSpec`.

Reply:
Yeah this is wrong, see the following:
```py
import torch
import torch.utils._cxx_pytree as pytree

@torch.compile(backend="eager", fullgraph=True)
def f(x, y):
    vals, spec = pytree.tree_flatten(x)
    return vals, spec, y.sin()

y = torch.randn(3)
x = [1, [2, [3, 4]]]
vals, spec, _ = f(x, y)
this_doesnt_work = pytree.tree_unflatten(vals, spec)
```

Reply:
I see. It may cause problems if we only compile part of the program.

There is an in-progress polyfill infra for C++ classes.

- #146678